engines/twp/actorswitcher.cpp

```diff
@@ -0,0 +1,165 @@
+	this->back = back_;
+	this->frame = frame_;
+	this->selectFunc = selectFunc_;
+	this->id = id_;
+}
+
+void ActorSwitcherSlot::select() {
+	if(selectFunc) {
+		selectFunc(id);
+	}
+}
+
+ActorSwitcher::ActorSwitcher() : Node("ActorSwitcher") {
+	_alpha = ENABLE_ALPHA;
+}
+
+Math::Matrix4 ActorSwitcher::transform(Math::Matrix4 trsf, int index) {
+	float animPos = _mouseOver ? _animPos : 1.f;
+	Math::Vector3d pos(SCREEN_WIDTH - MARGIN, SCREEN_HEIGHT - MARGIN - animPos * ACTOR_SEP * index, 0.f);
+	Math::Vector2d s(2.f, 2.f);
+	trsf.translate(pos);
+	Twp::scale(trsf, s);
+	return trsf;
+}
+
+float ActorSwitcher::getAlpha(size_t index) const {
+	if (index == (_slots.size() - 1))
+		return ENABLE_ALPHA;
+	if (_mode & asTemporaryUnselectable)
+		return INACTIVE_ALPHA;
+	if (_mode & asOn)
+		return _mouseOver ? ENABLE_ALPHA : _alpha;
+	return _mouseOver ? INACTIVE_ALPHA : DISABLE_ALPHA;
+}
+
+void ActorSwitcher::drawIcon(const Common::String &icon, Color backColor, Color frameColor, Math::Matrix4 trsf, int index) {
+	SpriteSheet *gameSheet = g_engine->_resManager.spriteSheet("GameSheet");
+	Texture *texture = g_engine->_resManager.texture(gameSheet->meta.image);
+	const SpriteSheetFrame &iconBackFrame = gameSheet->getFrame("icon_background");
+	const SpriteSheetFrame &iconActorFrame = gameSheet->getFrame(icon);
+	const SpriteSheetFrame &iconFrame = gameSheet->getFrame("icon_frame");
+	Math::Matrix4 t = transform(trsf, index);
+	float alpha = getAlpha(index);
+
+	drawSprite(iconBackFrame, texture, Color::withAlpha(backColor, alpha), t);
+	drawSprite(iconActorFrame, texture, Color::withAlpha(Color(), alpha), t);
+	drawSprite(iconFrame, texture, Color::withAlpha(frameColor, alpha), t);
+}
+
+void ActorSwitcher::drawCore(Math::Matrix4 trsf) {
+	if (_mouseOver) {
+		for (size_t i = 0; i < _slots.size(); i++) {
+			ActorSwitcherSlot &slot = _slots[i];
+			drawIcon(slot.icon, slot.back, slot.frame, trsf, i);
+		}
+	} else if (!_slots.empty()) {
+		ActorSwitcherSlot &slot = _slots[0];
+		drawIcon(slot.icon, slot.back, slot.frame, trsf, 0);
+	}
+}
+
+void ActorSwitcher::drawSprite(const SpriteSheetFrame &sf, Texture *texture, Color color, Math::Matrix4 trsf) {
+	Math::Vector3d pos(sf.spriteSourceSize.left - sf.sourceSize.getX() / 2.f, -sf.spriteSourceSize.height() - sf.spriteSourceSize.top + sf.sourceSize.getY() / 2.f, 0.f);
+	trsf.translate(pos);
+	g_engine->getGfx().drawSprite(sf.frame, *texture, color, trsf);
+}
+
+float ActorSwitcher::height() const {
+	float n = _mouseOver ? _slots.size() : 1;
+	return n * ACTOR_SEP;
+}
+
+int ActorSwitcher::iconIndex(Math::Vector2d pos) const {
+	float y = SCREEN_HEIGHT - pos.getY();
+	return _slots.size() - 1 - (int)((height() - y) / ACTOR_SEP);
+}
+
+Common::Rect ActorSwitcher::rect() const {
+	float h = height();
+	return Common::Rect(Common::Point(SCREEN_WIDTH - ACTOR_SEP, SCREEN_HEIGHT - h), ACTOR_SEP, h);
+}
+
+void ActorSwitcher::update(const Common::Array<ActorSwitcherSlot> &slots, float elapsed) {
+	if (_visible) {
```

Comment:
My personal taste in methods like this is remove one level of indentation and write the code like:
```C++
if (!_visible)
   return;

.... here goes the method body
```